### PR TITLE
Add the ability to consider A changed if B changes.

### PR DESCRIPTION
**Pull request by rsheeter**
Allows us to model dependencies between files. May need to become pattern-based. Baby steps.

### CODE REVIEW

File: fontir/src/depgraph.rs [hunk 1]
```diff
@@ -227,23 +306,41 @@ impl DepGraph {
...
+                // Add path and, recursively, things that depend on it
+                let mut to_add = Vec::new();
+                to_add.push(path);
+                while let Some(path) = to_add.pop() {
+                    if !changed.insert(path) {
+                        continue;
+                    }
+                    if let Some(deps) = self.dependencies.get(path) {
+                        deps.iter().for_each(|e| to_add.push(e));
+                    }
```
Comment (anthrotype) [inline fontir/src/depgraph.rs:330]: do we not need to check for cycles?

Comment (anthrotype) [reply fontir/src/depgraph.rs:330]: oh sorry that's exactly what what `if !changed.insert(path) continue;` does... 😅 
HashSet.insert returns false when it already contains the value

File: fontir/src/depgraph.rs [hunk 2]
```diff
@@ -227,23 +306,41 @@ impl DepGraph {
...
-                });
+                // Add path and, recursively, things that depend on it
+                let mut to_add = Vec::new();
+                to_add.push(path);
+                while let Some(path) = to_add.pop() {
+                    if !changed.insert(path) {
+                        continue;
+                    }
+                    if let Some(deps) = self.dependencies.get(path) {
+                        deps.iter().for_each(|e| to_add.push(e));
```
Comment (cmyr) [inline fontir/src/depgraph.rs:329]: I would write this as,
```suggestion
                        to_add.extend(deps.iter());
```

or maybe even something like,
```suggestion
                    to_add.extend(self.dependencies.get(path).into_iter().flat_map(Vec::iter));
```

although that might be a bit obscure. (Also I haven't checked that this compiles)

File: fontir/src/depgraph.rs [hunk 3]
```diff
@@ -343,12 +440,15 @@ mod tests {
         );
     }
 
-    fn one_changed_file_one_not(temp_dir: &TempDir) -> (PathBuf, PathBuf, DepGraph) {
-        let unchanged = temp_dir.path().join("a");
-        let changed = temp_dir.path().join("b");
+    fn write(temp_dir: &TempDir, path: &PathBuf, content: &str) -> PathBuf {
```
Comment (cmyr) [inline fontir/src/depgraph.rs:443]: `&Path`

**Review by cmyr — APPROVED** [review]
Did a quick pass just looking at Rust idioms & patterns, a few notes inline, nothing blocking.

File: fontir/src/depgraph.rs [hunk 4]
```diff
@@ -46,34 +67,52 @@ impl From<DepGraphSerdeRepr> for DepGraph {
...
-                        },
-                    }
-                })
-                .collect(),
+            entries: DepGraphEntriesSerdeRepr {
+                entries: dg
+                    .entries
+                    .iter()
+                    .map(|e| {
+                        let path = e.0.to_str().expect("Only UTF names please").to_string();
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/depgraph.rs:88]: There is no reason, will change as suggested.

[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/depgraph.rs:88]: Sadly this does not avoid alloc as we start with &PathBuf. I wanted a string to read/write to disk. We want to be able to have textual representation of those files and that's a great deal simpler if we have unicode content. So I take it back, there is a reason :)

File: fontir/src/depgraph.rs [hunk 5]
```diff
@@ -131,7 +171,11 @@ impl DepGraph {
         self.entries.contains_key(path)
     }
 
-    pub fn track(&mut self, path: &PathBuf, default_state: InitialState) -> Result<(), io::Error> {
+    fn track_internal(
+        &mut self,
+        path: &PathBuf,
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/depgraph.rs:176]: Ty. I don't know why, but I somehow thought PathBuf was what I should use.

[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/depgraph.rs:176]: Done

File: fontir/src/depgraph.rs [hunk 6]
```diff
@@ -157,6 +201,43 @@ impl DepGraph {
...
+            // Deps have to be tracked for this to work
+            self.track_internal(dep, default_state)?;
+
+            // And we need to remember the relationship
+            let dep = *dep;
+            if self.dependencies.contains_key(dep) {
+                self.dependencies.get_mut(dep).unwrap().push(path.clone());
+            } else {
+                self.dependencies.insert(dep.clone(), vec![path.clone()]);
+            }
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/depgraph.rs:235]: Nice, fixed.

File: fontir/src/depgraph.rs [hunk 7]
```diff
@@ -157,6 +201,43 @@ impl DepGraph {
...
+    /// Pay attention to path, we'd like to know if it changes.
+    ///
+    /// If path, or anything it depends on (recursive) changes then
+    /// path will be reported from [Self::changed()]. If the same path is
+    /// repeatedly tracked dependencies accumulate.
+    ///
+    /// All dependencies are themselves tracked and will report changes.
+    pub fn track_with_deps(
+        &mut self,
+        path: &PathBuf,
```
[the remark this answers is not in the document]
Comment (rsheeter) [reply fontir/src/depgraph.rs:218]: Done

File: fontir/src/depgraph.rs (hunk 2 above)
Comment (rsheeter) [reply fontir/src/depgraph.rs:329]: I went with `to_add.extend(deps.iter());`, terse and simple ftw

**Pull request merged**